### PR TITLE
Refactor DMMF generation and schema building away from deep recursion

### PR DESCRIPTION
**Pull request by tomhoule**
I tested this manually: it produces DMMF of the same size on the same schemas. The order of the objects is different, but I am  pretty sure the logic did not change, so it should be correct.

closes https://github.com/prisma/prisma/issues/3051

### CODE REVIEW

File: query-engine/query-engine/src/dmmf/schema/enum_renderer.rs [hunk 1]
```diff
@@ -21,13 +17,15 @@ impl<'a> Renderer<'a, ()> for DMMFEnumRenderer<'a> {
     }
 }
 
-impl<'a> DMMFEnumRenderer<'a> {
-    pub fn new(enum_type: &'a EnumType) -> DMMFEnumRenderer<'a> {
-        DMMFEnumRenderer { enum_type }
+impl<'a> DMMFEnumRenderer {
```
Comment (tomhoule) [inline query-engine/query-engine/src/dmmf/schema/enum_renderer.rs:20]: TODO: remove the lifetime param

File: query-engine/query-engine/src/dmmf/schema/type_renderer.rs [hunk 2]
```diff
@@ -1,183 +1,167 @@
-use super::*;
+use super::{DMMFTypeInfo, RenderContext, TypeKind};
+use query_core::{InputType, IntoArc, OutputType, ScalarType};
+
```
Comment (tomhoule) [inline query-engine/query-engine/src/dmmf/schema/type_renderer.rs:3]: The diff is in file is mostly from removing an indentation level.

File: query-engine/core/src/schema_builder/output_types/mutation_type.rs [hunk 3]
```diff
@@ -32,6 +35,51 @@ pub(crate) fn build(ctx: &mut BuilderContext) -> (OutputType, ObjectTypeStrongRe
...
+            append_opt(&mut fields, input_fields::nested_update_many_field(ctx, &rf));
+            append_opt(&mut fields, input_fields::nested_delete_many_field(ctx, &rf));
+            append_opt(&mut fields, input_fields::nested_upsert_field(ctx, &rf));
+
+            if feature_flags::get().connectOrCreate {
+                append_opt(&mut fields, input_fields::nested_connect_or_create_field(ctx, &rf));
+            }
+
+            input_object.set_fields(fields);
+        }
```
Comment (tomhoule) [inline query-engine/core/src/schema_builder/output_types/mutation_type.rs:76]: we could move the logic inside the for loops to the relevant modules instead free functions, it does not belong here

**Comment by tomhoule** [comment]
I'm going to merge this, since I'll be away next week. The CI/test suite work to reproducibly test for these stack overflows will happen in a follow-up PR: https://github.com/prisma/prisma-engines/pull/1081

**Comment by Jolg42** [comment]
@janpio Looks like this PR was merged and assigned to Milestone 2.5.0 I fixed it and put it in 2.6.0

**Pull request merged**